File: c_stuff/capi/argtypes/argtypes.c
#include<stdio.h>

void arg_int(int i)
{
    printf("arg_int: %d\n", i);
}

void arg_double(double d)
{
    printf("arg_double: %f\n", d);
}

void arg_string(char* s)
{
    printf("arg_string: "); 
    printf(s);
    printf("\n");    
}

void arg_list_int(int* l, int len)
{
    printf("arg_list_int: ");
    for (int i = 0; i < len; i++)
    {
        printf("%d ", l[i]);
    }
    printf("\n");
}

void arg_list_double(double* l, int len)
{
    printf("arg_list_double: ");
    for (int i = 0; i < len; i++)
    {
        printf("%1f ", l[i]);
    }
    printf("\n");
}

void arg_list_2d(int** l, int r, int c)
{
    printf("arg_list_2d:\n");
    for (int i = 0; i < r; i++)
    {
        for (int j = 0; j < c; j++)
        {
            printf("%d ", l[i][j]);
        }
        printf("\n");
    }
}

File: c_stuff/capi/argtypes/argtypes.py
import capi
import random

pth = capi.pydir("argtypes.dll")


@capi.cfunc
def arg_int(i, dll=pth): 
    pass
    
    
@capi.cfunc
def arg_double(d, dll=pth): 
    pass


@capi.cfunc
def arg_string(s, dll=pth): 
    pass
    
    
@capi.cfunc
def arg_list_int(l, i, dll=pth):
    pass
    
    
@capi.cfunc
def arg_list_double(l, i, dll=pth):
    pass
    
    
@capi.cfunc
def arg_list_2d(l, r, c, dll=pth):
    pass


arg_int(50)
arg_double(100.5)
arg_string("Test")
arg_list_int([1, 2, 3], 3)
arg_list_double([1.2, 3.4, 5.6, 7.8], 4)

n = 10
l = [[random.randint(100, 999) for i in range(n)] for j in range(n)]        
arg_list_2d(l, n, n)
